ddtrace/tracer/log.go

```diff
@@ -0,0 +1,109 @@
+	Service               string            `json:"service"`                 // Tracer Service
+	AgentURL              string            `json:"agent_url"`               // The address of the agent
+	AgentError            string            `json:"agent_error"`             // Any error that occurred trying to connect to agent
+	Debug                 bool              `json:"debug"`                   // Whether debug mode is enabled
+	AnalyticsEnabled      bool              `json:"analytics_enabled"`       // True if there is a global analytics rate set
+	SampleRate            string            `json:"sample_rate"`             // The default sampling rate for the rules sampler
+	SamplingRules         []SamplingRule    `json:"sampling_rules"`          // Rules used by the rules sampler
+	SamplingRulesError    string            `json:"sampling_rules_error"`    // Any errors that occurred while parsing sampling rules
+	Tags                  map[string]string `json:"tags"`                    // Global tags
+	RuntimeMetricsEnabled bool              `json:"runtime_metrics_enabled"` // Whether or not runtime metrics are enabled
+	HealthMetricsEnabled  bool              `json:"health_metrics_enabled"`  // Whether or not health metrics are enabled
+	ApplicationVersion    string            `json:"dd_version"`              // Version of the user's application
+	Architecture          string            `json:"architecture"`            // Architecture of host machine
+	GlobalService         string            `json:"global_service"`          // Global service string. If not-nil should be same as Service. (#614)
+}
+
+// agentReachable determines whether or not the tracer is able to connect
+// to endpoint, which should be the full URL returned by transport.endpoint
+func agentReachable(endpoint string) error {
+	req, err := http.NewRequest("POST", endpoint, nil)
+	if err != nil {
+		return fmt.Errorf("cannot create http request: %v", err)
+	}
+	_, err = defaultClient.Do(req)
+	if err != nil {
+		log.Warn("Unable to reach agent: %s", err)
+		return err
+	}
+	return nil
+}
+
+// logStartup generates a startupInfo for a tracer and writes it to the log in
+// JSON format.
+func logStartup(t *tracer) {
+	if !boolEnv("DD_TRACE_STARTUP_LOGS", true) {
+		return
+	}
+	tags := make(map[string]string)
+	for k, v := range t.globalTags {
+		tags[k] = fmt.Sprintf("%v", v)
+	}
+
+	info := startupInfo{
+		Date:                  time.Now().Format(time.RFC3339),
+		OSName:                osName(),
+		OSVersion:             osVersion(),
+		Version:               version.Tag,
+		Lang:                  "Go",
+		LangVersion:           runtime.Version(),
+		Env:                   t.config.env,
+		Service:               t.config.serviceName,
+		AgentURL:              t.transport.endpoint(),
+		Debug:                 t.config.debug,
+		AnalyticsEnabled:      !math.IsNaN(globalconfig.AnalyticsRate()),
+		SampleRate:            fmt.Sprintf("%f", t.rulesSampling.globalRate),
```

Comment:
Why isn't this field a float? Is there any issue with JSON encoding it?

Reply:
From the functional standpoint, this SHOULD be a float (or a number in the json terms: e.g. `0.7`).

Reply:
This cannot be a float, because floats may have `NaN` values, which are not json serializable, and will cause panics.

Reply:
No way to convert a `NaN` to `null` during serialization?

Reply:
As far as I'm concerned, string is fine. Unless this has to be a float for whatever parsing reasons.

Reply:
We *could* convert `NaN` to `nil`, but it's not as straight-forward as this, and the value seems dubious to me. Do we care that it's actually float type?

See: (https://github.com/golang/go/issues/3480)